### ISSUE TITLE
not cleaning all ibc files

### ISSUE THREAD

**Issue by archaeron** [idris-lang/Idris-dev#2710]
I'm trying to build a package, that I build with the last version.

It tells me: `This library was built with an earlier version of Idris. Please clean and rebuild.`
but after cleaning I get the same error.

This is on Windows.

```
λ idris --version
0.9.19.1
```

![idris_build](https://cloud.githubusercontent.com/assets/471043/10376569/c43d4240-6dfe-11e5-99d5-5b3e87a2070c.PNG)

**Comment by david-christiansen** [comment]
Are all modules listed in the .ipkg?

**Comment by jfdm** [comment]
more so, have you recently renamed files and modules since the last clean.

An alternate explanation is if you are using new versions of an installed library in which old ibc files exist.

**Comment by archaeron** [comment]
I don't have all the modules listed. But if it compiles, shouldn't it also clean?
Do I need to list all modules, even if I only want to export one?

I don't think I've renamed anything. And I have no dependencies.

**Comment by jfdm** [comment]
archaeron, idris cleans by using the the list of modules as a guide. If it ain't on the list it ain't getting scrubbed.

**Comment by archaeron** [comment]
It's just that the docs say

`idris --build maths.ipkg` will build all modules in the package
`idris --clean maths.ipkg` will delete all intermediate code and executable files generated when building.

so I assumed it cleans what it builds

**Comment by jfdm** [comment]
yes, and how do you tell idris what modules are in the package...

**Comment by david-christiansen** [comment]
I think the issue here may be that Idris will "helpfully" import packages that are in the source directory but not listed in the `.ipkg`, building an IBC in the process. This means that the build command can generate IBC files that will not be deleted by the clean command.

**Comment by david-christiansen** [comment]
This should be fixed!

**Comment by archaeron** [comment]
exactly, thanks a lot @david-christiansen :)

**Comment by david-christiansen** [comment]
I actually think this is a problem - in my opinion, the build command should fail if a module isn't listed as part of a package. This has bit me a few times in the past, and it should be fixed rather than being another thing for users to memorize!

**Comment by archaeron** [comment]
Ah, I understood "This should be fixed!" as: this is already fixed on master. I Agree, that it should be fixed if it isnt.

**Comment by jfdm** [comment]
I agree that this should be fixed at some point soon, this has rear-ended me too in the past. There are also a few other *known* problems encountered when using recently updated installed packages relating to stale `ibc` files. How we need do build and installation management needs to be revisited.
